### DISCUSSION TITLE
`RemoveDependency`: how to delete multiple things with same groupId?

### DISCUSSION THREAD

**Discussion by svaningelgem** [openrewrite/rewrite#4792] [Q&A] [answered]
I have this recipe:
```
recipeList:
  - org.openrewrite.maven.RemoveDependency:
      groupId: com.example.ws4
      artifactId: security
  - org.openrewrite.maven.RemoveDependency:
      groupId: com.example.ws4
      artifactId: functions
```

However, when ran I only see one is processed (the first):
```
[WARNING] Changes have been made to api-impl\pom.xml by:
[WARNING]     org.openrewrite.recommendations.CodeStyle
[WARNING]         org.openrewrite.xml.format.AutoFormat
[WARNING]         org.openrewrite.maven.RemoveDependency: {groupId=com.example.ws4, artifactId=security}
[WARNING]             org.openrewrite.maven.ChangeDependencyGroupIdAndArtifactId: {oldGroupId=com.example.ws4, oldArtifactId=*, newGroupId=com.example.ws5}
```

Why are the others not processed? (FYI: the other recipe also matches in that same pom file)
Or should I understand it as that the recipe is a set, with the key the class name?
=> that would make sense to remove duplicates from recipes -- but in this case it's not taking into account the parameters - which are different...
=> And if so: that would be a bug imho?

Oh, I also tried to extract them from the recipeList in their own recipe, and include that in the main list like this:
```
recipeList:
- com.example.removeSecurity
- com.example.removeFunctions
```

But that didn't work out either (also only security is removed)

**Comment by timtebeek** [comment]
Hi @svaningelgem ; I've quickly tried to replicate this issue with a unit test, but was not able to using this example:
```java
@Issue("https://github.com/openrewrite/rewrite/discussions/4792")
@Test
void removeTwoDependencies() {
    rewriteRun(
      spec -> spec.recipes(
        new RemoveDependency("junit", "junit", null),
        new RemoveDependency("com.google.guava", "guava", null)
      ),
      pomXml(
        """
          <project>
            <modelVersion>4.0.0</modelVersion>
            <groupId>com.mycompany.app</groupId>
            <artifactId>my-app</artifactId>
            <version>1</version>
            <dependencies>
              <dependency>
                <groupId>com.google.guava</groupId>
                <artifactId>guava</artifactId>
                <version>29.0-jre</version>
              </dependency>
              <dependency>
                <groupId>junit</groupId>
                <artifactId>junit</artifactId>
                <version>4.13.1</version>
                <scope>test</scope>
              </dependency>
            </dependencies>
          </project>
          """,
        """
          <project>
            <modelVersion>4.0.0</modelVersion>
            <groupId>com.mycompany.app</groupId>
            <artifactId>my-app</artifactId>
            <version>1</version>
          </project>
          """
      )
    );
}
```

Are you able to replicate this with a smaller sample perhaps?

As an implementation detail: the dependency is removed by calling out to a visitor to remove that specific tag:
https://github.com/openrewrite/rewrite/blob/094e714edea2e71c969467b26003c822d73af78c/rewrite-maven/src/main/java/org/openrewrite/maven/RemoveDependency.java#L83
If that tag can then not be found, perhaps because another recipe modified it in some way, then it will not be removed as it fails to match (different identity).
https://github.com/openrewrite/rewrite/blob/094e714edea2e71c969467b26003c822d73af78c/rewrite-xml/src/main/java/org/openrewrite/xml/RemoveContentVisitor.java#L37-L40

**Comment by svaningelgem** [comment]
That might be possible because I'm changing the groupId from `ws4` to `ws5`. I'll have a look tomorrow if removing that rule fixes it.
But that would mean that I need to duplicate the removal rules. Once for ws4 & once for ws5. Because the order is not guaranteed.

**Reply by shanman190** [reply, marked as answer]
Also to note, the groupId and artifactId I believe support glob patterns, so you could just throw a `?` (match single character only) or `*` (match all remaining characters) on the end of either value to get back down to a single recipe statement.

**Reply by svaningelgem** [reply]
Thanks, I'll try that, but for now I have [another issue](https://github.com/openrewrite/rewrite/issues/4796) with this removal. So I don't know if the `?/*` is actually working.

**Comment by madhura-23** [comment]
Hi 👋, I can help with this! 

To delete multiple dependencies with the same `groupId`, you can iterate over all matching dependencies and remove them programmatically. Here’s a general approach:

1. **Identify dependencies by `groupId`**:
   - Filter your dependency list to only include those with the target `groupId`.

2. **Iterate and remove**:
   ```java
   dependencies.stream()
               .filter(dep -> dep.getGroupId().equals("com.example"))
               .forEach(dep -> removeDependency(dep));
